src/DiffSharp.Core/Tensor.fs

```diff
@@ -1334,6 +1336,7 @@ and TensorOp =
     | AcosT of Tensor
     | AtanT of Tensor
     | NewT
+    | OpT of (* inputs *) Tensor list * (* gradients *) ((* tangent *) Tensor -> (* input-grad *) Tensor list)

```

Comment:
Would it perhaps be better to specialize unary and binary operations?

P.S. It's matter of taste, but I'd prefer DU labels or even better anonymous records instead of comments.

Reply:
Yes we can add DU labels - or probably a whole new type

Specialising unary/binary is ok but we'll need n-ary extensions eventually. Also ones producing multiple results though I'm guessing we'll always have "same-differentiation level" constraints on those.